File: aoc22_py/day_10.py
"""The solution to day 10."""
from advent_of_code_ocr import convert_array_6

from .aoc_helper import Solution


class Day(Solution):
    """The solution to day 10."""

    day = 10

    def __init__(self, raw: str):
        """Parse and store the raw data."""
        x = 1
        self.values: list[int] = []
        for line in raw.split("\n"):
            if not line.strip():
                continue
            self.values.append(x)
            if line != "noop":
                self.values.append(x)
                x += int(line.split()[1])

    def part_1(self) -> str | int | None:
        """Calculate the answer for part 1."""
        return sum(
            x_reg * (clock + 1)
            for clock, x_reg in enumerate(self.values)
            if clock % 40 == 19
        )

    def part_2(self) -> str | int | None:
        """Calculate the answer for part 2."""
        screen = [[False] * 40 for _ in range(6)]
        for clock, x_reg in enumerate(self.values):
            row, col = divmod(clock, 40)
            if -1 <= col - x_reg <= 1:
                screen[row][col] = True
        return convert_array_6(screen, fill_pixel=True, empty_pixel=False)


if __name__ == "__main__":
    Day.submit()

File: aoc22_rs/src/days/day10.rs
use crate::Day;

#[derive(Clone)]
pub struct Day10(Vec<isize>);

impl Day for Day10 {
    fn parse(input: &str) -> Self {
        let mut x = 1;
        let mut values = Vec::new();
        for line in input.lines() {
            values.push(x);
            if line != "noop" {
                values.push(x);
                x += line.split_once(' ').unwrap().1.parse::<isize>().unwrap();
            }
        }
        Self(values)
    }

    fn part1(&self) -> String {
        (0..6)
            .map(|i| 20 + i * 40)
            .map(|clock| clock as isize * self.0[clock - 1])
            .sum::<isize>()
            .to_string()
    }

    fn part2(&self) -> String {
        let mut screen = [[false; 40]; 6];
        for (clock, x_reg) in self.0.iter().enumerate() {
            let (row, col) = (clock / 40, clock % 40);
            let diff = col as isize - x_reg;
            if (-1..=1).contains(&diff) {
                screen[row][col] = true;
            }
        }
        let ascii_screen = screen
            .iter()
            .map(|row| {
                row.iter()
                    .map(|&x| if x { '#' } else { '.' })
                    .collect::<String>()
            })
            .collect::<Vec<_>>()
            .join("\n");
        advent_of_code_ocr::parse_string_to_letters(&ascii_screen)
    }
}
